sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -95,6 +95,7 @@ class SparkOptimizer(
       LimitPushDownThroughWindow,
       ConstantFolding,
       EliminateLimits),
+    Batch("Post Push down for V2 Relations", FixedPoint(1), PostV2ScanRelationPushDown),
```

Comment:
how does file source v1 overcome this issue?

Reply:
For FileSource v1, the issue is handled in [FileSourceStrategy](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L221), where the necessary columns are pruned and passed to `FileSourceScanExec`.